Add Node 10 runtime, add deprecation warning when deploying with Node 6, keep default to Node 6 but add warnings

thechenky opened the pull request:
<!--

Thank you for contributing to the Firebase community! Please fill out the form below.

Run the linter and test suite
==============================
Run `npm test` to make sure your changes compile properly and the tests all pass on your local machine. We've hooked up this repo with continuous integration to double check those things for you. 

-->


### Description
This PR does a couple things:
* adds Node 10 as a valid choice of runtime
* adds warnings printed on deployment when Node 6 is used since Node 6 is now deprecated.
* adds warnings printed on deployment when Node 10 is used as it's in Beta (this could be changed or removed - not sure if we should be using warnings for a Beta product)
* checks that firebase-functions SDK is at least v2.0.0 or higher to work with Node 8 and Node 10 runtimes, and prints warnings when it isn't with detailed instructions.
* refactors the logic to deduce runtime to make it easier to test. 

**NOTE: we are keeping the current behavior where the CLI will default to using Node 6 runtime when engines field is not specified. We will soon be requiring the engines field for all deployments, and failing if engines field is not set. Because of this I have added large warnings when Node 6 is chosen in the event where engine is not specified in `package.json`. The plan is to completely remove this default behavior under a new major version release in ~1 month after releasing this change (under a minor release).**
	 
### Scenarios Tested
* Unit tests for `getRuntimeChoice` for deploys using Node 6, Node 8, and Node 10
* Ran a deploy of a test project using different runtimes specified in package.json:

No engines specified in package.json, default runtime gets set to Node 6 by the CLI:
![image](https://user-images.githubusercontent.com/31747099/56942051-3d40af00-6acd-11e9-916f-b8a6c69687d5.png)

Node 6 deployment output:
```
✔  functions: Finished running predeploy script.
i  functions: ensuring necessary APIs are enabled...
✔  functions: all necessary APIs are enabled
i  functions: preparing functions directory for uploading...
⚠  functions: Deploying functions to Node 6 runtime, which is deprecated. Node 8 is available and is the recommended runtime.
i  functions: packaged functions (32.9 KB) for uploading
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 6 function helloWorld(us-central1)...
✔  functions[helloWorld(us-central1)]: Successful update operation.

✔  Deploy complete!
```

Node 8 deployment output:
```
✔  functions: Finished running predeploy script.
i  functions: ensuring necessary APIs are enabled...
✔  functions: all necessary APIs are enabled
i  functions: preparing functions directory for uploading...
i  functions: packaged functions (32.9 KB) for uploading
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 8 function helloWorld(us-central1)...
✔  functions[helloWorld(us-central1)]: Successful update operation.

✔  Deploy complete!
```
Node 10 deployment output (no warnings for Beta):
```
✔  functions: Finished running predeploy script.
i  functions: ensuring necessary APIs are enabled...
✔  functions: all necessary APIs are enabled
i  functions: preparing functions directory for uploading...
i  functions: packaged functions (32.9 KB) for uploading
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 10 function helloWorld(us-central1)...
✔  functions[helloWorld(us-central1)]: Successful update operation.

✔  Deploy complete!
```

Node 8 deployment with old firebase-functions SDK version:
```
✔  functions: Finished running predeploy script.
i  functions: ensuring necessary APIs are enabled...
✔  functions: all necessary APIs are enabled
i  functions: preparing functions directory for uploading...
⚠  functions: You must have a firebase-functions version that is at least 2.0.0. Please run npm i --save firebase-functions@latest in the functions folder.
i  functions: packaged functions (32.9 KB) for uploading
✔  functions: functions folder uploaded successfully
i  functions: updating Node.js 10 function helloWorld(us-central1)...
```

* Ran the integration tests in firebase-functions SDK: `./integration_test/run_tests.sh chenky-test-3` - all tests passing.

GitTom commented:
> it adds warnings printed on deployment when Node 10 is used as it's in Beta (this could be changed or removed - not sure if we should be using warnings for a Beta product)

The constant warnings that I'm using an unsupported (beta) runtime are annoying.  If necessary I think this warning should be moved to somewhere where the developer will see it a more reasonable amount (eg. < 20 times).

As it is now, with active development, I've probably seen it thousands of times.

mbleigh commented:
Let's just remove the beta warning. If we want to do anything let's just
put (beta) after the runtime name in the deploy line

On Fri, Apr 26, 2019, 6:57 AM GitTom <notifications@github.com> wrote:

> it adds warnings printed on deployment when Node 10 is used as it's in
> Beta (this could be changed or removed - not sure if we should be using
> warnings for a Beta product)
>
> The constant warnings that I'm using an unsupported (beta) runtime are
> annoying. If necessary I think this warning should be moved to somewhere
> where the developer will see it a more reasonable amount (eg. < 20 times).
>
> As it is now, with active development, I've probably seen it thousands of
> times.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/firebase/firebase-tools/pull/1223#issuecomment-487066246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAAH7RZC6BRHOWINJDCCR3PSMC3DANCNFSM4HITKZFQ>
> .
>

thechenky commented on src/getRuntimeChoice.ts:
```diff
@@ -0,0 +1,82 @@
+function _handleNode10(): string {
+  utils.logWarning(
+    clc.bold.yellow("functions: ") +
+      "Deploying functions to Node 10 runtime which is in Beta. " +
```
Removed

thechenky commented on src/functionsDeployHelper.js:
```diff
@@ -215,43 +215,14 @@ function getRuntimeName(runtime) {
   return runtime;
 }
 
+// TODO: default behavior is no longer supported, so remove this
```
Removed this too, good catch

thechenky commented:
Thanks @GitTom for chiming in - I've now removed the warnings for Node 10 deploys.

GitTom commented:
> Thanks @GitTom for chiming in - I've now removed the warnings for Node 10 deploys.

Great, thanks @thechenky.  Much appreciated.

mbleigh commented on src/getRuntimeChoice.ts:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  if (nodeVersion("6")) {
+    return _handleNode6();
```
I'm not a big fan of relying on function hoisting (defining functions below where you use them), and each of these functions are really quite small. WDYT about something streamlined:

```ts
const ENGINE_RUNTIMES = {
  "6": "nodejs6",
  "8": "nodejs8",
  "10": "nodejs10",
};

const choice = _.get(loaded, "engines.node");
const runtime = ENGINE_RUNTIMES[choice];
if (!runtime) {
  // throw unsupported engine error
}

if (runtime === "nodejs6") {
 // print deprecation warning
}

return runtime
```

This doesn't include the functions SDK check (sorry rushing a bit here) but I think is a little more straightforward as an approach.

thechenky replied:
Agreed - once I removed the warnings from Node 10 there's definitely not much happening in these functions. Will refactor.

bkendall commented on src/test/getRuntimeChoice.spec.ts:
```diff
@@ -0,0 +1,71 @@
+import * as sinonchai from "sinon-chai";
+import * as utils from "../utils";
+import { getRuntimeChoice } from "../getRuntimeChoice";
+chai.use(sinonchai);
```
I didn't see this before - you don't need to do this. It's covered by the mocha bootstrap: https://github.com/firebase/firebase-tools/blob/master/src/test/helpers/mocha-bootstrap.js

thechenky replied:
I need to use sinonchai though, which is different from chaiAsPromised - it breaks the tests when I remove this line.

bkendall replied:
😱 I didn't realize that was a thing! I thought it was `chai-as-promised`. I'm going to remember this exists!

mbleigh replied:
Should we move this to the bootstrap anyway, then it's available everywhere?

thechenky replied:
I'm not sure if that's possible - it's either chaiAsPromised or sinon, and some tests will need one while others would need the other. I don't think you can use both, can you?

thechenky replied:
Oh, actually looks like I can do this! Made the changes in the last commit.

mbleigh commented on src/getRuntimeChoice.ts:
```diff
@@ -10,67 +10,69 @@ import * as utils from "./utils";
+  10: "nodejs10",
+};
+
+export const ENGINES_FIELD_REQUIRED_MSG = clc.bold(
```
Looks like you're exporting these so you can use them in tests. In general if you're doing something just so you can test it's a kind of smell...

I think the `to.throw()` is pretty loose in how it looks at error messages (simple includes or regex or something). You could maybe not export these and just do something like `to.throw(FirebaseError, "unsupported")`?

thechenky replied:
Yeah, these are exported for tests - this makes it so the error message is all in one place, so if the message itself changes, the tests will not break since the message is defined in one place. So, if something changes in the string to modify "unsupported" to, say, "not supported", then the tests would break even though functionality has remained the same.

That being said, I'm not super happy with the structure of this code and tests - I feel like these functions should all be tested separately and should not be throwing or logging warnings, but should be returning something sensible that the test can then check, but so little is happening here that it's not worth creating all these extra wrapper functions just to make it more modularized. So, I'm rather torn.

mbleigh replied:
I'd lean toward not exporting even if it makes the test brittle against the language. If we change the language, the test will fail and we'll have to change it but that's very straightforward. Exporting the error messages just for testing ends up feeling like an overkill solution for a quickly fixed problem.

thechenky replied:
Went with keeping the constants https://github.com/firebase/firebase-tools/pull/1223#discussion_r280012336. But if there are strong feelings against, I don't mind changing.

thechenky commented on src/getRuntimeChoice.ts:
```diff
@@ -10,67 +10,69 @@ import * as utils from "./utils";
 var cjson = require("cjson");
 
+const ENGINE_RUNTIMES: any = {
+  6: "nodejs6",
```
Yeah, I had this in quotes at first but prettier got angry and thew `Unnecessarily quoted property '10' found.` errors. So then I was forced to remove them :(

thechenky commented:
Quick recap of resolved points:
* Removed beta warning on Node 10 deployment
* Added chai.use(sinon) to bootstrap code and removed from test
* Streamlined getRuntimeChoice function per @mbleigh's suggestion
* Refactored getRuntimeName to streamline in a similar way to getRuntimeChoice
* Prettier does not allow "unnecessary quotes" on properties
* Added warning/error message verification to tests (jury still out whether this has made them better)

Unresolved:
* Tests using exported constants for warning strings
* Add changelog.txt entry for Node 10 runtime

bkendall commented on src/test/helpers/mocha-bootstrap.js:
```diff
@@ -1,7 +1,9 @@
+const sinonChai = require("sinon-chai");
 
 chai.use(chaiAsPromised);
+chai.use(sinonChai);
```
![giphy (5)](https://user-images.githubusercontent.com/160236/56974904-9be54780-6b24-11e9-8dbf-ee7e91cf2032.gif)

thechenky replied:
bahahaha :)

bkendall commented on src/test/runtimeChoiceSelector.spec.ts:
```diff
@@ -0,0 +1,102 @@
+});
+
+describe("getRuntimeChoice", () => {
+  const sandbox: sinon.SinonSandbox = sinon.createSandbox();
```
```suggestion
  const sandbox = sinon.createSandbox();
```

Pretty sure it's redundant to declare the type inline like that :)

thechenky replied:
Oh, I thought this was part of Typescript, like, defining the type of the constant. If it redundant in all cases or just this in particular?

bkendall replied:
I don't think it's necessary. (a) you can't re-assign a `const` anyway, and (b) if it's a `let`, the compiler will shout if you've given it a type once and then try to change it.

[typescript playground](https://www.typescriptlang.org/play/#src=let%20n%3A%20number%3B%0D%0A%0D%0An%20%3D%203%3B%0D%0A%0D%0An%20%3D%20%224%22%3B%0D%0A%0D%0Aconst%20s%20%3D%20%224%22%0D%0A%0D%0As%20%3D%20%22asdf%22%3B%0D%0A%0D%0Alet%20v%20%3D%204%3B%0D%0A%0D%0Av%20%3D%20%22foo%22)

thechenky replied:
Done.

mbleigh commented on src/functionsDeployHelper.js:
```diff
@@ -208,50 +208,13 @@ function pollDeploys(operations, printSuccess, printFail, printTooManyOps, proje
-    options.config.path(options.config.get("functions.source")),
-    "package.json"
+  utils.logWarning(
+    clc.bold.red(
```
Let's do yellow, since logWarning will automatically put a little yellow :warning: emoji

thechenky replied:
I wanted to really grab attention since this will be a breaking change in the future, although I can see that red is used to indicate errors and so might confuse the user into thinking something is wrong (something *is* wrong, though, that they should fix). These are the two options, I'm fine going with the yellow, defer to you.

Red:
![image](https://user-images.githubusercontent.com/31747099/57008536-b1e71c80-6ba5-11e9-8c41-3b93d8ac6604.png)

Yellow:
![image](https://user-images.githubusercontent.com/31747099/57008581-068a9780-6ba6-11e9-9cef-d5a113532484.png)

mbleigh replied:
Let's go with yellow, but maybe add newlines around it, so:

logger.info()
utils.logWarning(...)
logger.info()

I also think maybe we should add in something like:

```
For example:

{
  {"engines": {"node": "6"}}
}
```

That way it's really clear what they need to do.

thechenky replied:
Changed to this:
![image](https://user-images.githubusercontent.com/31747099/57028868-0ebdf380-6bf5-11e9-8436-c251b4682077.png)

mbleigh replied:
LGTM but how about a little less indentation :)

thechenky replied:
lololol fair, fair.

thechenky replied:
![image](https://user-images.githubusercontent.com/31747099/57032710-e38cd180-6bff-11e9-83ee-334d02c58fcf.png)

bkendall commented on src/runtimeChoiceSelector.ts:
```diff
@@ -0,0 +1,95 @@
+  "Deploying functions to Node 6 runtime, which is deprecated. Node 8 is available " +
+  "and is the recommended runtime.";
+
+export const FUNCTIONS_SDK_VERSION_TOO_OLD_WARNING =
```
I don't think any of these constant strings should be exported - there's nothing outside this function that would need them, right?

(I see the tests, I'm excluding those in this thought and making a comment down there)

bkendall replied:
(I actually may change my mind on this. I'll leave it up to you - I sent you a internal link out-of-band.)

thechenky replied:
Discussed offline - according to internal unit test best practices documentation:

*Though increasing visibility should be avoided for methods, it's generally fine on constants to help keep tests in sync as those constants change. It can also be valuable on fields representing UI elements, which are visible to the user and part of a class's contract, but not normally exposed via its API.*

I'd like to keep the constants exported, unless there are strong concerns.

mbleigh commented on src/runtimeChoiceSelector.ts:
```diff
@@ -0,0 +1,95 @@
+// tslint:disable-next-line
+var cjson = require("cjson");
+
+const MESSAGE_FRIENDLY_RUNTIMES: any = {
```
You could make the types on these `{[key: string]: string}` for slightly more complete typing.

thechenky replied:
Oh love that, was looking for that, thanks!

bkendall commented on src/test/runtimeChoiceSelector.spec.ts:
```diff
@@ -0,0 +1,102 @@
+    cjsonStub.returns({ engines: { node: "6" }, dependencies: { "firebase-functions": "2.0.0" } });
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.deep.equal("nodejs6");
+    expect(warningSpy).calledWith(runtime.DEPRECATION_WARNING_MSG);
```
Alright, it's been a while since I've done this, but here's a different way to test this than copying/exporting entire strings and making sure they match.

Each of the deprecation warnings has some specific text in them. You can use a regular expression to match those keywords (around all the other content). To do this with `calledWith`, sinon provides [matchers](https://sinonjs.org/releases/latest/matchers/). It would make this check look like:

```suggestion
    expect(warningSpy).calledWith(sinon.match(/Node 6 runtime.+deprecated/));
```

How does that feel?

thechenky replied:
Leaving the constants in as per our other offline conversation.

bkendall replied:
👍

bkendall commented on src/test/runtimeChoiceSelector.spec.ts:
```diff
@@ -0,0 +1,102 @@
+    expect(warningSpy).not.called;
+  });
+
+  // TODO: Add this test back in when we remove runtime default behavior
```
How are we going to find this again? Can you make an internal bug and put `TODO(b/xxxxxx)` here so we can find/track it (same below and any other appropriate place)

thechenky replied:
Good call - annotated all TODOs with the correct bug number.

thechenky commented on package.json:
```diff
@@ -62,6 +62,7 @@
     ]
   },
   "dependencies": {
+    "@types/semver": "^6.0.0",
```
This is used in a non-test file (runtimeChoiceSelector) to check if Functions SDK version is lower than 2.0.0.

bkendall replied:
`semver` is, yes. But the `@types` package is only used by the typescript compiler. Once the code is transformed, the `@types` packages are no longer needed (and so they only need to be in `devDependencies`)

thechenky replied:
Ooooh, did not know that, thanks for letting me know!

thechenky commented:
Ugh how do I remove that googlebot thing, does it not know that @bkendall is a googler?
